### ISSUE TITLE
Undefined Sanitizer emits alignment requirement of 8 bytes for double inside struct on i386

### ISSUE THREAD

**Issue by ADKaster** [llvm/llvm-project#54845]
The following program generates an -fsanitize=alignment error with clang++-14.0.1 but not gcc-11.2.0

Is this code actually undefined behavior? Or is the clang code in clang::CodeGen::CodeGenFunction::EmitCXXNewExpr  improperly emitting the wrong sanitizer metadata? I've seen this same error with more complex code that tries to create the equivalent of std::aligned_storage using `alignof(T)` and being surprised that alignof(double) is actually 4 on i386 but UBSAN is complaining that my storage is supposed to have 8 byte alignment.

aligntest.cpp
```cpp
#include <new>
 
struct S {
    short b;
    double c;
};
 
__attribute__((noinline)) void foo(double& d) {
    new (&d) double(0);
}
 
int main() {
    S s {};
    foo(s.c);
}
```

clang++-14 output:

```console
~/dev$ clang++ --version
Ubuntu clang version 14.0.1-++20220407063131+ec13fed5867f-1~exp1~20220407063214.111
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
~/dev$ clang++ -m32 -Og -ggdb3 -fsanitize=undefined aligntest.cpp
~/dev$ ./a.out
aligntest.cpp:9:14: runtime error: constructor call on misaligned address 0xfffc2a7c for type 'double', which requires 8 byte alignment
0xfffc2a7c: note: pointer points here
  00 00 00 00 00 00 00 00  00 00 00 00 98 19 f0 f7  00 30 bc f7 ca c4 9b f7  01 00 00 00 54 2b fc ff
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior aligntest.cpp:9:14 in 
```

g++-11 output:
```console
~/dev$ g++ --version
g++ (Ubuntu 11.2.0-7ubuntu2) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

~/dev$ g++ -m32 -Og -ggdb3 -fsanitize=undefined aligntest.cpp
~/dev$ ./a.out
~/dev$ 
```

The i386 psABI has an alignment requirement of 4 bytes for the double here, as shown by pahole

```console
$ pahole ./a.out 
struct S {
	short                      b;                    /*     0     2 */

	/* XXX 2 bytes hole, try to pack */

	double                     c;                    /*     4     8 */

	/* size: 12, cachelines: 1, members: 2 */
	/* sum members: 10, holes: 1, sum holes: 2 */
	/* last cacheline: 12 bytes */
};
```

**Comment by BertalanD** [comment]
Candidate patch: https://reviews.llvm.org/D124736
